flocx_market/api.py

```diff
@@ -0,0 +1,25 @@
+import os
+
+from flask import Flask, request
+from flask_restful import Api
+from flask_migrate import Migrate
+import pymysql
+from dotenv import load_dotenv
+from db import db
+
+from resources.offer import Offer, OfferList
+
+app = Flask(__name__)
+load_dotenv(".env")
+app.config['SQLALCHEMY_DATABASE_URI'] = os.environ.get('DATABASE_URI')
+app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
+app.config['PROPAGATE_EXCEPTIONS'] = True
+api = Api(app)
+migrate = Migrate(app, db)
```

Comment:
Starting up the API should not install the database tables (if I'm understanding this code correctly); they should be completely separate. You can see how Ironic or esi-leap does it, but perhaps it's fine for now, because the "OpenStack" way of doing things require the work that Danni and Leo are doing.

Reply:
hey, the database tables are only installed with migrations, starting it up won't initialize any tables. This is just configuration for the ORM. Does that answer your concern?

Reply:
It's possible I don't fully understand how Flask works! Does it require knowledge of the database?

Reply:
the sqlalchemy track modifications if not specified will be set to TRUE and Flask will use it's own tracker which is poorer than SQL Alchemy's and that's why I set it to False.
For the propagate exceptions, I guess it's not that important since debug is already true:

"""PROPAGATE_EXCEPTIONS
Exceptions are re-raised rather than being handled by the app’s error handlers. If not set, this is implicitly true if TESTING or DEBUG is enabled.
Default: None"""


The orm does require some knowledge of the db such as how to access it

Reply:
Is there any way to separate the API from the database? Most OpenStack projects I've seen have three layers: database, business object, and REST API, with all the business logic encompassed in the business object. This allows for, say, a manager service to manipulate objects without going through the REST API (useful for expiring objects that have moved past a certain date).

Reply:
I'm also concerned about the db part inside api.py. I have nowhere to start with oslo db configuration. I guess we need a db folder with a db api.

Reply:
yes we can separate them. However, the db piece will still need the Flask app to initialize the SQL Alchemy ORM @tzumainn

Reply:
Is the "Flask app" the webapp that runs the REST API? If so, I think there's something problematic if the database layer requires that to work. Or I could be misunderstanding... ?

Reply:
The flask app needs to initialize the ORM, but it shouldn't need to call any sort of database migration. The api code shouldn't know about alembic nor should it require the `flask_migrate` module. This should all be handled by a separate tool.